File: BloodPressure/src/main/java/org/caojun/bloodpressure/preference/SeekBarPreference.java
package org.caojun.bloodpressure.preference;

import android.content.Context;
import android.content.DialogInterface;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

/**
 * Created by CaoJun on 2017/7/26.
 */

public class SeekBarPreference extends DialogPreference implements DialogInterface.OnClickListener, SeekBar.OnSeekBarChangeListener {

    private static final String ANDROIDNS = "http://schemas.android.com/apk/res/android";
    private SeekBar mSeekBar;
    private TextView mValueText;
    private Context mContext;
    private int mDefault = 24, mMax = 24, mValue = 24;

    public SeekBarPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
        mDefault = attrs.getAttributeIntValue(ANDROIDNS,"defaultValue", 24);
        mMax = attrs.getAttributeIntValue(ANDROIDNS,"max", 24);
    }

    @Override
    protected View onCreateDialogView() {
        LinearLayout.LayoutParams params;
        LinearLayout layout = new LinearLayout(mContext);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(6,6,6,6);

        mValueText = new TextView(mContext);
        mValueText.setGravity(Gravity.CENTER_HORIZONTAL);
        mValueText.setTextSize(32);
        params = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.FILL_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
        layout.addView(mValueText, params);

        mSeekBar = new SeekBar(mContext);
        mSeekBar.setOnSeekBarChangeListener(this);
        layout.addView(mSeekBar, new LinearLayout.LayoutParams(LinearLayout.LayoutParams.FILL_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));

        if (shouldPersist()) {
            mValue = getPersistedInt(mDefault);
        }

        mSeekBar.setMax(mMax);
        mSeekBar.setProgress(mValue);
        return layout;
    }

    @Override
    protected void onBindDialogView(View v) {
        super.onBindDialogView(v);
        mSeekBar.setMax(mMax);
        mSeekBar.setProgress(mValue);
    }

    @Override
    protected void onSetInitialValue(boolean restore, Object defaultValue)
    {
        super.onSetInitialValue(restore, defaultValue);
        if (restore) {
            mValue = shouldPersist() ? getPersistedInt(mDefault) : 0;
        } else {
            mValue = (Integer) defaultValue;
        }
    }

    @Override
    public void onProgressChanged(SeekBar seek, int value, boolean fromTouch) {
        String t = String.valueOf(value);
        mValueText.setText(t);
        callChangeListener(new Integer(value));
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {

    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {

    }

    public void setMax(int max) {
        mMax = max;
    }

    public int getMax() {
        return mMax;
    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        switch (which) {
            case DialogInterface.BUTTON_POSITIVE:
                if (shouldPersist()) {
                    persistInt(mSeekBar.getProgress());
                }
                break;
        }
        super.onClick(dialog, which);
    }
}

File: WidgetLibrary/src/main/java/org/caojun/widget/SeekBarPreference.kt
package org.caojun.widget

import android.content.Context
import android.content.DialogInterface
import android.preference.DialogPreference
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.TextView

/**
 * Created by CaoJun on 2017-12-8.
 */
class SeekBarPreference: DialogPreference, DialogInterface.OnClickListener, SeekBar.OnSeekBarChangeListener {

    private val ANDROIDNS = "http://schemas.android.com/apk/res/android"
    private var mSeekBar: SeekBar? = null
    private var mValueText: TextView? = null
    private var mContext: Context? = null
    private var mDefault = 24
    private var mMax = 24
    private var mValue = 24

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        mContext = context
        mDefault = attrs?.getAttributeIntValue(ANDROIDNS, "defaultValue", 24)?:24
        mMax = attrs?.getAttributeIntValue(ANDROIDNS, "max", 24)?:24
    }

    override fun onCreateDialogView(): View {
        val layout = LinearLayout(mContext)
        layout.orientation = LinearLayout.VERTICAL
        layout.setPadding(6, 6, 6, 6)

        mValueText = TextView(mContext)
        mValueText?.gravity = Gravity.CENTER_HORIZONTAL
        mValueText?.textSize = 32f
        val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        layout.addView(mValueText, params)

        mSeekBar = SeekBar(mContext)
        mSeekBar?.setOnSeekBarChangeListener(this)
        layout.addView(mSeekBar, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT))

        if (shouldPersist()) {
            mValue = getPersistedInt(mDefault)
        }

        mSeekBar?.max = mMax
        mSeekBar?.progress = mValue
        return layout
    }

    override fun onBindDialogView(v: View) {
        super.onBindDialogView(v)
        mSeekBar?.max = mMax
        mSeekBar?.progress = mValue
    }

    override fun onSetInitialValue(restore: Boolean, defaultValue: Any) {
        super.onSetInitialValue(restore, defaultValue)
        mValue = if (restore) {
            if (shouldPersist()) getPersistedInt(mDefault) else 0
        } else {
            defaultValue as Int
        }
    }

    override fun onProgressChanged(seek: SeekBar, value: Int, fromTouch: Boolean) {
        mValueText?.text = value.toString()
        callChangeListener(value)
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {

    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {

    }

    override fun onClick(dialog: DialogInterface, which: Int) {
        when (which) {
            DialogInterface.BUTTON_POSITIVE -> if (shouldPersist()) {
                persistInt(mSeekBar?.progress?:0)
            }
        }
        super.onClick(dialog, which)
    }
}